geom/twkb_write.go

```diff
@@ -0,0 +1,534 @@
+package geom
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+)
+
+// MarshalTWKB accepts a geometry and generates the corresponding TWKB byte slice.
+func MarshalTWKB(geom Geometry,
```

Comment:
Regarding the function signature, a few things stick out to me:

- The user shouldn't have to supply `hasZ` and `hasM`. Instead, `MarshalTWKB` should be able to get this info using `geom.CoordinatesType().Is3D()` and `geom.CoordinatesType().IsMeasured()`.
- `precXY` should always be specified, but `precZ` and `precM` only make sense for 3D and measured geometries.
- My understanding of the spec is that writers _shouldn't_ close rings. So I don't think we need to make that behaviour configurable. Does that match your understanding?
- `hasSize` and `hasBBox` seem sensible to allow to be configured, although IMO ideally we would default these to both `false`.
- `idList` is sensible to allow, some (many?) users wouldn't need it.

What do you think about using the functional options pattern? The signature could look something like this:

```
func MarshalTWKB(g Geometry, precXY int, opts ...TWKBWriterOption) ([]byte, error) {
    var os twkbWriterOptionSet
    os.precZ = precXY // default if not set by opt
    os.precY = precXY // default if not set by opt
    for _, opt := range opts {
        opt(&os)
    }
    ... // the rest of the logic
}

type TWKBWriterOption func(*twkbWriterOptionSet)

type twkbWriterOptionSet struct {
    precZ, precM int
    hasSize, hasBBox bool
    idList []int64
}

func TWKBPrecisionZ(int precZ) TWKBWriterOption {
    return func(s *twkbWriterOptionSet) {
        s.precZ = precZ
    }
}

... // the rest of the options

```

Reply:
Regarding your comments about the encoder:
* Yes, I'll rework the function to allow options. I'd considered it, but figured I'd wait to see what approach you'd prefer.
* Agreed that many users won't use the ID list. It's there for adherence to the spec.
* Agreed that many flags can often be omitted or defaulted to false.
* Technically, supplying `hasZ` and `hasM` separately from the `Geometry` could provide additional functionality of stripping those coordinate values during encoding if they are `false` (although that's not the way the code is currently written). If you're happy to always deduce these from the geometry, those parameters can be omitted as you mentioned.
* You're correct that the spec says the encoder shouldn't close rings. The reason I added that capability was to allow byte-for-byte comparison against an actual PostGIS implementation which does seem to close rings. If possible I'd like to retain the capability, but hide it behind an option.

Reply:
Sounds good!